metaflow/plugins/kfp/kfp.py

```diff
@@ -1,86 +1,293 @@
+    return step_op
+
+def pre_start_container_op():
+    """
+    Container op that corresponds to the 'pre-start' step of Metaflow.
+
+    Note: The public docker image is a copy of the internal docker image we were using (borrowed from aip-kfp-example).
+    """
+
+    pre_start_op = kfp.components.func_to_container_op(pre_start_op_func, base_image='ssreejith3/mf_on_kfp:python-curl-git')
+    return pre_start_op
+
+def create_command_templates_from_graph(graph):
+    """
+    Create a map of steps to their corresponding command templates. These command templates help define the command
+    to be used to run that particular step with placeholders for the `run_id` and `datastore_root` (location of the datastore).
+
+    # Note:
+    # Level-order traversal is adopted to keep the task-ids in line with what happens during a local metaflow execution.
+    # It is not entirely necessary to keep this order of task-ids if we are able to point to the correct input-paths for
+    # each step. But, using this ordering does keep the organization of data more understandable and natural (i.e., `start`
+    # gets a task id of 1, next step gets a task id of 2 and so on with 'end' step having the highest task id.
+    """
+
+    """
+    Returns the python command template to be used for each step.
+    
+    This method returns a string with placeholders for `datastore_root` and `run_id`
+    which get populated based on the outputs of our initial setup (i.e., when these values are known) in 
+    the pipeline.
+    The rest of the command string is populated using the passed arguments which are known before the run starts.
+    
+    An example constructed command template (to run a step named `hello`):
+    "python downloaded_flow.py --datastore s3 --datastore-root {ds_root} " \
+                     "step hello --run-id {run_id} --task-id 2 " \
+                     "--input-paths {run_id}/start/1"
+    """
+    def build_cmd_template(step_name, task_id, input_paths):
+        python_cmd = "python {downloaded_file_name} --datastore s3 --datastore-root {{ds_root}} " \
+                     "step {step_name} --run-id {{run_id}} --task-id {task_id} " \
+                     "--input-paths {input_paths}".format(downloaded_file_name=DEFAULT_DOWNLOADED_FLOW_FILENAME,
+                                                            step_name=step_name, task_id=task_id, input_paths=input_paths)
+        return python_cmd
+
+    steps_deque = deque(['start']) # deque to process the DAG in level order
+    cur_task_id = 0
+
+    # set of seen steps, i.e., added to the queue for processing
+    seen_steps = set(['start'])
+    # Mapping of steps to task ids
+    task_id_map = {}
+    # Mapping of steps to their command templates
+    command_template_map = {}
+
+    while len(steps_deque) > 0:
+        cur_step = steps_deque.popleft()
+        cur_task_id += 1
+        task_id_map[cur_step] = cur_task_id
+        cur_node = graph.nodes[cur_step]
```

Comment:
style: current_step instead of cur_step.
I suggest to lean towards easier to read, rather than even a bit more verbose.

Reply:
Done